schematic/base.py

```diff
@@ -0,0 +1,446 @@
+        while nlr:
+            nlr = nlr.next_less_restrictive
+            depth += 1
+        return depth
+
+    def value_is_compatible(self, value):
+        """Checks to see if the given value can be inserted into a column of
+           the type described by this instance.
+        Args:
+          value: The value to check, a string
+        Returns:
+          True if _value_is_compatible_superset(self, value) is True and
+          this type isn't parameterized
+        Raises:
+          NotImplementedError: This should be implemented by subclasses
+                               if subclasses are parameterized.
+        """
+        if self.parameterized:
+            raise NotImplementedError
+        else:
+            return self._value_is_compatible_superset(value)
+
+    def _value_is_compatible_superset(self, value):
+        """Checks to see if the given value can be inserted into a column of
+           the group of types described by this class.
+           This is used, for example, to check if a value could
+           fit into any VARCHAR class in a SQL database, whereas
+           value_is_compatible would be used
+           to check if a value could fit into specifically
+           into a VARCHAR(256) class.
+        Args:
+          value: The value to check, a string
+        Raises:
+          NotImplementedError: This should be implemented by subclasses.
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    def get_parameter_for_value(value):
+        """Get the parameter for a parameterized implementation
+           of this class that is required to fit the given value.
+
+        Args:
+          value: the value to return the parameter for.
+        Raises:
+          NotImplementedError: Subclasses should implement this.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def from_value(cls, value):
+        """Create an instance of this class that is compatible with value.
+
+        Args:
+          value: The value to return an instance for
+        Returns:
+          An instance of cls that can fit the value.
+        Raises:
+          NoCompatibleParameterError: if value cannot fit in any instance of this class.
+        """
+        if cls.parameterized:
+            if not cls()._value_is_compatible_superset(value):
+                raise ValueError(
+                    "Value {} not compatible with any instance of {}".format(
+                        value, cls.name))
+            else:
+                return cls(parameter=cls.get_parameter_for_value(value))
+        else:
+            return cls()
+
+
+class TableColumn(ABC, DictableMixin, NameSqlMixin):
+    """DB-agnostic base class for storing info about a column in a table.
+
+    Attributes:
+      name: The identifier for the column
+      type: A TableColumnType instance with type information for the column.
+    """
+
+    def __init__(self, name, column_type):
+        self.name = name
```

Comment:
```suggestion
        self.name = name.upper()
```
since columns are treated as type-insensitive in TableDefinition, we might as well nip this in the bud and not worry as much about case in TableDefinition